src/bsp/platform/esp32/xi_bsp_io_fs_esp32.c

```diff
@@ -0,0 +1,307 @@
+/* Copyright (c) 2003-2017, LogMeIn, Inc. All rights reserved.
+ *
+ * This is part of the Xively C Client library,
+ * it is licensed under the BSD 3-Clause license.
+ */
+
+#include <sys/stat.h>
+#include <errno.h>
+#include <assert.h>
+#include <string.h>
+#include <stdio.h>
+
+#include <xi_bsp_debug.h>
+#include <xi_bsp_io_fs.h>
+#include <xi_bsp_fwu.h>
+#include <xi_bsp_mem.h>
+
+#include "esp_vfs.h"
+#include "esp_vfs_fat.h"
+#include "esp_ota_ops.h"
+
+#define XI_ESP32_MAX_FILENAME_LEN 64
+#define ESP32_FS_BASE_PATH "/spiflash"
+
+#define XI_BSP_IO_FS_CHECK_CND( cnd, e, s )                                              \
+    if ( ( cnd ) )                                                                       \
+    {                                                                                    \
+        ( s ) = ( e );                                                                   \
+        perror( "Xi FS BSP errno" );                                                     \
+        goto err_handling;                                                               \
+    }
+
+/* These functions are implemented in the application layer to report status */
+extern void esp32_xibsp_notify_update_started( const char* filename, size_t file_size );
+extern void esp32_xibsp_notify_chunk_written( size_t chunk_size, size_t offset );
+
+/* The size of the buffer to be used for reads. */
+const size_t xi_bsp_io_fs_buffer_size = 1024;
+/* ESP32 Over The Air FW updates API handle */
+static esp_ota_handle_t open_firmware_bin_handle = 0;
+
+/* translates bsp errno errors to the xi_bsp_io_fs_state_t values */
+xi_bsp_io_fs_state_t xi_bsp_io_fs_posix_errno_2_xi_bsp_io_fs_state( int errno_value )
+{
+    xi_bsp_io_fs_state_t ret = XI_BSP_IO_FS_STATE_OK;
+
+    switch ( errno_value )
+    {
+        case EBADF:
+        case EACCES:
+        case EFAULT:
+        case ENOENT:
+        case ENOTDIR:
+            ret = XI_BSP_IO_FS_RESOURCE_NOT_AVAILABLE;
+            break;
+        case ELOOP:
+        case EOVERFLOW:
+            ret = XI_BSP_IO_FS_ERROR;
+            break;
+        case EINVAL:
+            ret = XI_BSP_IO_FS_INVALID_PARAMETER;
+            break;
+        case ENAMETOOLONG:
+            ret = XI_BSP_IO_FS_INVALID_PARAMETER;
+            break;
+        case ENOMEM:
+            ret = XI_BSP_IO_FS_OUT_OF_MEMORY;
+            break;
+        default:
+            ret = XI_BSP_IO_FS_ERROR;
+    }
+
+    return ret;
+}
+
+xi_bsp_io_fs_state_t
+xi_bsp_io_fs_open( const char* const resource_name,
+                   const size_t size,
+                   const xi_bsp_io_fs_open_flags_t open_flags,
+                   xi_bsp_io_fs_resource_handle_t* resource_handle_out )
+{
+    char filepath[XI_ESP32_MAX_FILENAME_LEN] = "";
```

Comment:
isn't = { 0 }; better?

Reply:
I usually go with `""` for consistency (all strings, empty or not, are initialized using `"`).

But if our style guideline is to use `= { 0 };`, I certainly prefer being consistent at that level. I'll change it.